### ISSUE TITLE
Assign extra topics during rebalance

### ISSUE THREAD

**Issue by alistairking** [twmb/franz-go#220]
Hey!

Thanks for the really nice library.
I'm in the process of trying to migrate some code from confluent_kafka_go (librdkafka) over to use franz-go, but I've run into a problem with altering topic assignments during rebalance.

It's a little unorthodox, but this consumer subscribes to a regex, and adds extra topics to the assignment based on those that it gets assigned.
So for example, it might subscribe to `^.+\.foo$`, get assigned `topicA.foo`, and then during the rebalance, it adds and assignment for `topicA.bar` alongside `topicA.foo`.

In rdkafka this works fine, but it seems franz-go is too smart and skips these topics:
```
2022-10-11T19:13:42Z ERR BUG! consumer was assigned topic that we did not ask for in ConsumeTopics nor ConsumePartitions, skipping!
```
https://github.com/twmb/franz-go/blob/3191842a81033342e8d37a529bd0a1b3d190fd9f/pkg/kgo/consumer.go#L924

Basically, I want to be able to subscribe to one set of topics, and ensure that for each assigned topic, the same instance consumes from another related topic.

The only idea I have to doing this in franz-go is to run a second Client that I use for "ad-hoc" subscriptions and then whenever a topic gets assigned in the regex client, I use [AddConsumeTopics](https://github.com/twmb/franz-go/blob/v1.8.0/pkg/kgo/consumer.go#L719) to add a subscription for the companion topics in the second Client.
It'll require a bunch more coordination logic, but I think it's doable and shouldn't add _too_ much overhead to the kafka cluster (I hope).

Any ideas?

**Comment by twmb** [comment]
Did you implement a custom balancer?

I'm not sure how this would work in the context of incremental rebalancing (KIP-429). I was staring a bit at the code yesterday and I think wiring this in would be a bit complicated to support doing this in one client alone. I'm guessing, but does each topic have a single partition as well?

Having a second client seems like a decent solution. In client 1 (regex consumer), you can use `OnPartitionsAdded` to call `AddConsumeTopics` in client 2, and you can use `OnPartitionsRevoked` to call `PurgeTopicsFromClient` in client 2. I _think_ this should work...

I'm not sure how much sense to support this within one client, and it'd be a bit awkward for cooperative consuming. I'm not sure how librdkafka supports this in addition to cooperative consuming. This is easier to support for eager consuming: all clients just give up all partitions and then rejoin. For cooperative, clients tell the leader which partitions they are currently consuming ... and I'm not sure how to handle the extra side topics, both for the join request and for when to revoke these extra partitions within the client.

**Comment by twmb** [comment]
An alternative would be to implement a custom balancer and consume from all possible topics: `^.+\.(?:foo|bar)$`. In the leader, balance only the foo topics, and then have a second pass that gives all bar topics to the member that is assigned the corresponding foo topic.

There are a bunch of functions in this package to help writing your own balancer. It's a bit tricky due to trying to keep the API generic true to the underlying protocol, but then also provide helpers for the common (`"consumer"`) protocols.

**Comment by alistairking** [comment]
Thanks for the quick reply.

Yeah, each topic does have a single partition. I probably should have been more clear about that -- I guess more accurately when a consumer instance gets assigned `topicA.foo-0`, it also wants `topicA.bar-0`.

I don't use a custom balancer, so I'm not sure how it works under the hood. To be fair, I am using the round-robin balancer (switching to cooperative-sticky is actually one of the reasons I'm porting this code over to franz due to a bug in the Go bindings for rdkfaka: https://github.com/confluentinc/confluent-kafka-go/issues/649), so I'm not 100% sure that my approach would work with a sticky balancer in rdkafka.

A custom balancer sounds like a pretty good idea, but I think it'd mean a bit more refactoring that I care for in the short term. I might give a second client a try and see how that goes.

Thanks for the help.

**Comment by alistairking** [comment]
Ah, just ran into the first couple road blocks with the dual client approach.

One is that I can't create an empty client: https://github.com/twmb/franz-go/blob/ac130a166ab5d4751af60902b9a84fc5f5cbaf0f/pkg/kgo/config.go#L347-L350

I hacked around this for now by subscribing to a non-existent topic, which logs a bunch of errors (that I can't quite tell if they're causing problems or just noise):
```
INF metadata update triggered why="re-updating due to inner errors: UNKNOWN_TOPIC_OR_PARTITION{nonexistent-topic}
```
This causes other problems because not all of the "bar" topics exist. With rdkafka I got the warning in-band and was able to log it once and then swallow any further messages.

I guess I'll give the custom balancer approach a try.

**Comment by twmb** [comment]
A couple things:

How were you doing this with librdkafka _without_ a custom balancer? I'm not sure how / where you added the extra topics. Also, how did you remove the extra topics from being consumed once a rebalance happened? What balancer were you using with librdkafka?

If an API existed to allow _adding_ extra topics after group assignment has happened, what do you think about an extra requirement that you must also say which extra topics are revoked once a revoke happens?

```go
// AdjustGroupBalancePlan allows adjusting a group balance plan after the
// leader has assigned partitions but before consuming of those partitions
// happens. The leader's plan is passed to onAssign; any extra topic partitions
// returned will have their offsets fetched.
//
// If using a cooperative balancer, any extra partitions you add are not
// revoked on rebalance. onRevoked is given partitions that are being revoked,
// you must return any extra partitions that should be revoked. onRevoke is not
// used for eager rebalancers; eager rebalancers always revoke everything.
func AdjustGroupBalancePlan(onAssign, onRevoke func(map[string][]int32) map[string][]int32) GroupOpt
```

I don't know how feasible this option is, but I think that'd solve your problem.

re: can't create an empty client, I _could_ remove this restriction. The restriction made sense before I added `AddConsumeTopics`. However, I'm still not sure it's worth removing this restriction because I'm not sure it's useful to allow joining a group with no topics. I didn't realize that the `bar` topics you mention are _also_ in the same group, and _also_ have commits. That makes it a bit challenging as well.

Custom balancer is also an option -- overall, you're already somewhat doing custom balancing, but you're piggybacking on existing balancers and then modifying the end results. I'm not sure how technically sound that is -- you'll have problems if you switch to another client that doesn't support the behavior your need -- but I guess if all group members consistently follow the same behavior, it may be fine.

**Comment by alistairking** [comment]
I just passed a rebalance callback to [`Consumer.Subscribe`](https://github.com/confluentinc/confluent-kafka-go/blob/master/kafka/consumer.go#L64), 
and then in the rebalance callback I added the extra TopicPartitions that I wanted, and called [`Consumer.Assign`](https://github.com/confluentinc/confluent-kafka-go/blob/master/kafka/consumer.go#L105) with the full set of partitions this consumer instance wanted to handle ¯\_(ツ)_/¯

I was using the round-robin balancer, so it's possible that this is why everything Just Worked -- on rebalance all the partitions were revoked?

That AdjustGroupBalancePlan API _seems_ like it would work.
I managed to get a custom balancer working today, and I think this is basically what I'm doing. I created a balancer that is a wrapper around another balancer, strips out the "sibling" topics, lets the child balancer balance the primary topics, and then goes through the plan and adds sibling topics to the members that have the primary topics (basically what you suggested originally). I've only done lightweight smoke testing so far, but from what I can tell, it seems to work!

At the moment I'm still wrapping the round-robin balancer, but once I confirm this works I'll try switching to cooperative-sticky.

**Comment by ssorren** [comment]
Maybe off topic @twmb but I have an incremental rebalancer implementation in kgo as part of https://github.com/aws/go-kafka-event-source/tree/main/streams


Built on jog, don’t know if satisfies the KIP or not.  I doubt it

**Comment by twmb** [comment]
@alistairking since you're able to make this work, I think I'm going to leave the `AdjustGroupBalancePlan` option as a feature request for now (good first issue, help wanted). It's a pretty niche request that I'm not the most keen to wire in right now. I'll happily review a pull request if you have one :sweat_smile: otherwise I'll leave it on the back burner similar to my own requested feature #214.

@ssorren that's quite an impressive usage of the library and building your own balancer. It almost certainly doesn't satisfy a KIP because, well, every KIP defined balancer is already in the library :D. I'm curious to see where you're taking it.

**Comment by ssorren** [comment]
> @ssorren that's quite an impressive usage of the library and building your own balancer. It almost certainly doesn't satisfy a KIP because, well, every KIP defined balancer is already in the library :D. I'm curious to see where you're taking it.

Well, the library I built provides for StateStores and the idea is to ensure that a state store is fully consumed/initialized on the consumer that will be receiving the partition before assigning it. Next step will be to have consumers assigned "stand by" partitions, so that we can recover form failures quickly.

**Comment by twmb** [comment]
I spent a solid chunk of time working on this to see how feasible AdjustAddLossFn would be to add. In short, it's a bit impractical to add to the client itself. Adding new topics is fine, but once a topic is added to the client, the client deems the topic _interesting_ and will continue expressing interest in the topic once a rebalance happens. So, if the client is assigned "foo" and adds "bar", then both "foo" and "bar" will be topics to consume on the next rebalance. The client would have to purge "bar" once a rebalance happens.

This is all a bit risky because there's a separate flow where you could manually add and remove those topics, and there would have to be a fair chunk of instrumentation to add just to determine "does this topic exist because we were explicitly assigned the topic, or does the topic exist because it is a sibling topic?".

I think, since a manual balancer approach works somewhat, I'm going to close this request for now. It's more challenging than I expected with some ambiguous edge conditions. Perhaps, if a similar request comes up down the line, I'll revisit.

**Comment by alistairking** [comment]
I really appreciate the time you spent looking at this and all the advice! 

Feel free to close the issue, I'm pretty happy with the custom balancer approach.
